src/Ztm.WebApi/SqlCallbackRepository.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Newtonsoft.Json;
+using Ztm.Data.Entity.Contexts;
+using Ztm.Data.Entity.Contexts.Main;
+
+namespace Ztm.WebApi
+{
+    public class SqlCallbackRepository : ICallbackRepository
+    {
+        readonly IMainDatabaseFactory db;
+
+        public SqlCallbackRepository(IMainDatabaseFactory db)
+        {
+            if (db == null)
+            {
+                throw new ArgumentNullException(nameof(db));
+            }
+
+            this.db = db;
+        }
+
+        public async Task<Callback> AddAsync(IPAddress registeringIp, Uri url, CancellationToken cancellationToken)
+        {
+            if (registeringIp == null)
+            {
+                throw new ArgumentNullException(nameof(registeringIp));
+            }
+
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+
+            using (var db = this.db.CreateDbContext())
+            {
+                var callback = await db.WebApiCallbacks.AddAsync(new WebApiCallback()
+                {
+                    Id = Guid.NewGuid(),
+                    RegisteredIp = registeringIp,
+                    RegisteredTime = DateTime.UtcNow,
+                    Url = url,
+                }, cancellationToken);
+
+                await db.SaveChangesAsync(cancellationToken);
+
+                return ToDomain(callback.Entity);
+            }
+        }
+
+        public async Task SetCompletedAsyc(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            using (var dbtx = await db.Database.BeginTransactionAsync(IsolationLevel.Serializable, cancellationToken))
+            {
+                var update = await db.WebApiCallbacks.FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                if (update == null)
+                {
+                    throw new KeyNotFoundException($"Id {id} is not found");
```

Comment:
I think we should not throw if it cannot find row to update. Just ignore instead. The reason is this operation on some database is very expensive (e.g. for distributed database).